src/main/kotlin/io/arrowkt/service/AccountService.kt

```diff
@@ -0,0 +1,31 @@
+package io.arrowkt.service
+
+import arrow.core.Option
+import java.math.BigDecimal
+import java.time.LocalDate
+
+enum class AccountType {
+    CHECKING,
+    SAVINGS
+}
+
+interface AccountService<Account, Amount, Balance> {
+
+    fun open(
+        no: String,
+        name: String,
+        rate: Option<BigDecimal>,
+        openingDate: Option<LocalDate>,
+        accountType: AccountType
+    ): AccountOperation<Account>
+
+    fun close(no: String, closeDate: Option<LocalDate>): AccountOperation<Account>
+
+    fun debit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun credit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun balance(no: String): AccountOperation<Balance>
+
+    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> = TODO()
```

Comment:
I could not figure out how to use `binding` here... 

The result should be:
```
def transfer(from: String, to: String, amount: Amount): AccountOperation[(Account, Account)] = for { 
    a <- debit(from, amount)
    b <- credit(to, amount)
  } yield ((a, b))
```

Reply:
Oh god monster monad stack. I believe 

`typealias AccountOperation<A> = Kleisli<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, A>`

That's what's supposed to be, and then you can get `ReaderT.monad(EitherT.monad(IO.monad()))` and that may get you `binding`.

Reply:
Time for me to learn Haskell....

Reply:
Nah we've moved past this. This builds up to something better.

Reply:
if I do
`typealias AccountOperation<A> = Kleisli<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, A>`

and

```
    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> =
            binding<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, Pair<Account, Account>>(ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad()))) {
                val a = debit(from, amount).bind()
                val b = credit(to, amount).bind()
                Pair(a, b)
            }.fix()
```

I get
`/home/coka/IdeaProjects/frdomain.kt/src/main/kotlin/io/arrowkt/service/AccountService.kt: (42, 114): Type mismatch: inferred type is KleisliMonad<EitherTPartialOf<ForIO, AccountServiceException> /* = Kind<Kind<ForEitherT, ForIO>, AccountServiceException> */, AccountRepository> but Monad<EitherTPartialOf<ForIO, AccountServiceException> /* = Kind<Kind<ForEitherT, ForIO>, AccountServiceException> */> was expected`

Reply:
start with

```
ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad())).binding {

}
```

and then see what's telling you about `debit` and `credit`. You're in the right direction, just keep on pushing!

Reply:
thanks @pakoito it compiles now.... and I have no idea how

Reply:
You were trying to pull a top-level function and IJ failed to decide which one.